### ISSUE TITLE
[prefer-function-type] suggested fix breaks mergeable interface

### ISSUE THREAD

**Issue by tadhgmister** [typescript-eslint/typescript-eslint#2438]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

[Playground](https://typescript-eslint.io/play#ts=5.3.2&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgOoWFAJgYTgZxQG8AoZZACgEoAuZfMKUAcwG4SBfEhAGwPzQZseQslLkEAexAMoAVwRhJUamLLlkUCGDlQQyEBADuyAApRJADwCeFMAAtg%2BADRqNGuAAdPPW2DhQzNquDk4AgoGuCHA8PBHM%2BLT0jCxu7hpaOnrIAORYcP7IQYZQBRBYyEb2EPrRsalw%2BqAMjUg57OnIHM7q5BxUHV2cJEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wlq9hMy%2BSgHsmtIqXRgA2uGw5e0UdEgKIAXQUBfEDqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false)

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-function-type": ["error"]
  }
}
```

```TS
interface WeirdCase {
  (): string;
}
class WeirdCase {
  constructor() {
    return new Proxy(this, {
      apply(target, thisArg, callArgs): string {
        return 'data generated when calling an instance';
      },
    });
  }
}
```

**Expected Result**

an interface with the same name as a class in the same scope will merge namespaces and therefore be considered valid. This is the correct way to specify to typescript that instances of my class are in fact callable.

**Actual Result**

the `prefer-function-type` rule suggests replacing the interface with a `type` declaration which is semantically invalid since it no longer merges with class and instead clashes with it.

**Additional Info**

issue also exists for `no-empty-interface` with default `{"allowSingleExtends": false}`
```ts
type A = ()=>string;
interface WeirdCase extends A {}
class WeirdCase { }
```
auto fix creates semantically invalid code.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.10.1` |
| `@typescript-eslint/parser`        | `3.10.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.2.0` |
| `node`                             | `v12.13.1` |

**Comment by tadhgmister** [comment]
I'd be interested in picking this up, if someone could point me to any reference for how to use `RuleContext` with regards to inspecting variables vs types etc that'd be a big help! 😄

**Comment by bradzacher** [comment]
This will have to wait for 4.0 to merge for proper scope analysis.

Docs on scope analysis APIs are here: https://eslint.org/docs/developer-guide/scope-manager-interface#top

**Comment by tadhgmister** [comment]
I just took another look and realized correct behaviour is a little more complicated than I originally realized, adding the info here for my future self or other implementor:

1. empty interface with another interface or class the fix is to just remove it
2. empty interface with extends with another interface can move the extends to that definition and remove the empty one
     - if there are more than one other interface definition which do we move it to? defining multiple empty interfaces meaning there are multiple interface to remove makes this not totally trivial
3. empty interface with extends with class also defined is allowed, indicates the class has implicit implementation.

and for `prefer-function-type`

1. single call signature with other interfaces present leads to similar issue as `2` above.
2. single call signature and `extends Function` with a class present does actually indicate the class has function methods like `bind`, `call` etc, so it does in fact add value.  Feels wrong to just allow this though, trying to explain it in docs even in an example is daunting.

**Comment by olafur-palsson** [comment]
I've run into this one a couple of times. Best solution I've found is to simply turn `no-empty-interface` off to avoid auto-fixing issues.
```ts
type ConfigType = typeof config; // config is a '{ ...something } as const' variable
export interface AppConfig extends ConfigType {}
export class AppConfig {
   // some logic
}  
```
Auto-fixing results in
```ts
type ConfigType = typeof config;
export type AppConfig = ConfigType
export class AppConfig {
```
Which results in `TS2300: Duplicate identifier 'AppConfig'.`

Can I recommend turning of the auto-fixing until this issue is fixed? The current fix is assuming that `type` and `interface` are interchangable, but they're clearly not since TS 4.0.

**Comment by bradzacher** [comment]
You really should avoid declaration merging classes. It's a really unsafe code pattern.

```ts
interface Foo {
  prop: string;
}
class Foo {}

const x = new Foo();

x.prop.length
// oops runtime crash TS was unable to catch

class Bar {
  prop: string;
// Property 'prop' has no initializer and is not definitely assigned in the constructor. ✅✅✅
}
```

**Comment by Josh-Cena** [comment]
> issue also exists for `no-empty-interface` with default `{"allowSingleExtends": false}`
> 
> ```ts
> type A = ()=>string;
> interface WeirdCase extends A {}
> class WeirdCase { }
> ```

I am taking `no-empty-interface` off the scope of this issue, because that would be practically a duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/1640, and is IMO the reason why there is `allowSingleExtends` in the first place

**Comment by bradzacher** [comment]
I'm going to close this bug out. It's been opened for 3 years and has gathered just 2 reactions.
It's clearly not a case that the community encounters at all - which makes sense given my comment above about it being an unsafe anti-pattern to do this.
Additionally this style of API of having a thing that is both a `new`able class and a callable function is not a common pattern today with the advent of ES6 classes.

I'm going to call this a known edge-case that we shan't be fixing.
